modules/code-editor/src/backend/editor.ts

```diff
@@ -0,0 +1,130 @@
+import * as sdk from 'botpress/sdk'
+import fs from 'fs'
+import _ from 'lodash'
+import path from 'path'
+
+import { Config } from '../config'
+
+import { EditableFile, FileType, TypingDefinitions } from './typings'
+
+const FILENAME_REGEX = /^[0-9a-zA-Z_\-.]+$/
+
+export default class Editor {
+  private bp: typeof sdk
+  private _botId: string
+  private _typings: TypingDefinitions
+  private _config: Config
+
+  constructor(bp: typeof sdk, botId: string, config: Config) {
+    this.bp = bp
+    this._botId = botId
+    this._config = config
+  }
+
+  async fetchFiles() {
+    return {
+      actionsGlobal: this._config.allowGlobal && (await this._loadFiles('/actions', 'action')),
+      actionsBot: await this._loadFiles('/actions', 'action', this._botId)
+    }
+  }
+
+  async _validateContent({ name, botId, type }: Partial<EditableFile>) {
+    if (!botId || !botId.length) {
+      if (!this._config.allowGlobal) {
+        throw new Error(`Global files are restricted, please check your configuration`)
+      }
+    } else {
+      if (botId !== this._botId) {
+        throw new Error(`Please switch to the correct bot to change its actions.`)
+      }
+    }
+
+    if (type !== 'action') {
+      throw new Error('Invalid file type Only actions are allowed at the moment')
+    }
+
+    if (!FILENAME_REGEX.test(name)) {
+      throw new Error('Filename has invalid characters')
+    }
+  }
+
+  async saveFile(file: EditableFile): Promise<void> {
+    this._validateContent(file)
+    const { location, botId, content } = file
+    const ghost = botId ? this.bp.ghost.forBot(this._botId) : this.bp.ghost.forGlobal()
+
+    return ghost.upsertFile('/actions', location, content)
+  }
+
+  async loadTypings() {
+    if (this._typings) {
+      return this._typings
+    }
+
+    const sdkTyping = fs.readFileSync(path.join(__dirname, '/../botpress.d.js'), 'utf-8')
```

Comment:
this is more a question than a comment. what are .d.js files ? I imagine it simply is a transpiled version of .d.ts, how does this work in practice, how is it used ?

Reply:
The build.extras.js file processing in the module builder convert ts files into js.

All I needed was the content of the botpress.d.ts file, so the fact that it changed the extension isn't important. I'm only reading the content of the file, to return it to monaco